// @sequelize/core/lib/dialects/abstract/query-interface-typescript.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var query_interface_typescript_exports = {};
__export(query_interface_typescript_exports, {
  AbstractQueryInterfaceTypeScript: () => AbstractQueryInterfaceTypeScript
});
module.exports = __toCommonJS(query_interface_typescript_exports);
var import_node_assert = __toESM(require("node:assert"));
var import_isEmpty = __toESM(require("lodash/isEmpty"));
var import_deferrable = require("../../deferrable");
var import_errors = require("../../errors");
var import_model_internals = require("../../model-internals.js");
var import_query_types = require("../../query-types");
var import_deprecations = require("../../utils/deprecations");
var import_query_interface_internal = require("./query-interface-internal.js");
class AbstractQueryInterfaceTypeScript {
  sequelize;
  queryGenerator;
  #internalQueryInterface;
  /**
   * @param sequelize The sequelize instance.
   * @param queryGenerator The query generator of the dialect used by the current Sequelize instance.
   * @param internalQueryInterface The internal query interface to use.
   *                               Defaults to a new instance of {@link AbstractQueryInterfaceInternal}.
   *                               Your dialect may replace this with a custom implementation.
   */
  constructor(sequelize, queryGenerator, internalQueryInterface) {
    this.sequelize = sequelize;
    this.queryGenerator = queryGenerator;
    this.#internalQueryInterface = internalQueryInterface ?? new import_query_interface_internal.AbstractQueryInterfaceInternal(sequelize, queryGenerator);
  }
  /**
   * Returns the database version.
   *
   * @param options Query Options
   */
  async fetchDatabaseVersion(options) {
    const payload = await this.#internalQueryInterface.fetchDatabaseVersionRaw(options);
    (0, import_node_assert.default)(payload.version != null, "Expected the version query to produce an object that includes a `version` property.");
    return payload.version;
  }
  /**
   * Create a new database schema.
   *
   * **Note:** We define schemas as a namespace that can contain tables.
   * In mysql and mariadb, this command will create what they call a database.
   *
   * @param schema Name of the schema
   * @param options
   */
  async createSchema(schema, options) {
    const sql = this.queryGenerator.createSchemaQuery(schema, options);
    await this.sequelize.queryRaw(sql, options);
  }
  /**
   * Drop a single schema
   *
   * **Note:** We define schemas as a namespace that can contain tables.
   * In mysql and mariadb, this command will create what they call a database.
   *
   * @param schema Name of the schema
   * @param options
   */
  async dropSchema(schema, options) {
    const dropSchemaQuery = this.queryGenerator.dropSchemaQuery(schema);
    let sql;
    let queryRawOptions;
    if (typeof dropSchemaQuery === "string") {
      sql = dropSchemaQuery;
      queryRawOptions = options;
    } else {
      sql = dropSchemaQuery.query;
      queryRawOptions = { ...options, bind: dropSchemaQuery.bind };
    }
    await this.sequelize.queryRaw(sql, queryRawOptions);
  }
  /**
   * Show all defined schemas
   *
   * **Note:** this is a schema in the [postgres sense of the word](http://www.postgresql.org/docs/9.1/static/ddl-schemas.html),
   * not a database table. In mysql and mariadb, this will show all databases.
   *
   * @param options
   *
   * @returns list of schemas
   */
  async showAllSchemas(options) {
    const showSchemasSql = this.queryGenerator.listSchemasQuery(options);
    const schemaNames = await this.sequelize.queryRaw(showSchemasSql, {
      ...options,
      raw: true,
      type: import_query_types.QueryTypes.SELECT
    });
    return schemaNames.map((schemaName) => schemaName.schema);
  }
  /**
   * Drop a table from database
   *
   * @param tableName Table name to drop
   * @param options   Query options
   */
  async dropTable(tableName, options) {
    const sql = this.queryGenerator.dropTableQuery(tableName, options);
    await this.sequelize.queryRaw(sql, options);
  }
  /**
   * Drop all tables
   *
   * @param options
   */
  async dropAllTables(options) {
    const skip = (options == null ? void 0 : options.skip) || [];
    const allTables = await this.showAllTables(options);
    const tableNames = allTables.filter((tableName) => !skip.includes(tableName.tableName));
    const dropOptions = { ...options };
    if (this.sequelize.dialect.supports.dropTable.cascade && dropOptions.cascade === void 0) {
      dropOptions.cascade = true;
    }
    for (const tableName of tableNames) {
      const foreignKeys = await this.showConstraints(tableName, { ...options, constraintType: "FOREIGN KEY" });
      await Promise.all(foreignKeys.map(async (fk) => this.removeConstraint(tableName, fk.constraintName, options)));
    }
    for (const tableName of tableNames) {
      await this.dropTable(tableName, dropOptions);
    }
  }
  /**
   * Show all tables.
   *
   * @param options
   */
  async showAllTables(options) {
    const sql = this.queryGenerator.listTablesQuery(options);
    return this.sequelize.queryRaw(sql, { ...options, raw: true, type: import_query_types.QueryTypes.SELECT });
  }
  /**
   * Returns a promise that will resolve to true if the table or model exists in the database, false otherwise.
   *
   * @param tableName - The name of the table or model
   * @param options - Query options
   */
  async tableExists(tableName, options) {
    const sql = this.queryGenerator.tableExistsQuery(tableName);
    const out = await this.sequelize.query(sql, { ...options, type: import_query_types.QueryTypes.SELECT });
    return out.length === 1;
  }
  /**
   * Describe a table structure
   *
   * This method returns an array of hashes containing information about all attributes in the table.
   *
   * ```js
   * {
   *    name: {
   *      type:         'VARCHAR(255)', // this will be 'CHARACTER VARYING' for pg!
   *      allowNull:    true,
   *      defaultValue: null
   *    },
   *    isBetaMember: {
   *      type:         'TINYINT(1)', // this will be 'BOOLEAN' for pg!
   *      allowNull:    false,
   *      defaultValue: false
   *    }
   * }
   * ```
   *
   * @param tableName
   * @param options Query options
   */
  async describeTable(tableName, options) {
    var _a;
    const table = this.queryGenerator.extractTableDetails(tableName);
    if (typeof options === "string") {
      (0, import_deprecations.noSchemaParameter)();
      table.schema = options;
    }
    if (typeof options === "object" && options !== null) {
      if (options.schema) {
        (0, import_deprecations.noSchemaParameter)();
        table.schema = options.schema;
      }
      if (options.schemaDelimiter) {
        (0, import_deprecations.noSchemaDelimiterParameter)();
        table.delimiter = options.schemaDelimiter;
      }
    }
    const sql = this.queryGenerator.describeTableQuery(table);
    const queryOptions = { ...options, type: import_query_types.QueryTypes.DESCRIBE };
    try {
      const data = await this.sequelize.queryRaw(sql, queryOptions);
      if ((0, import_isEmpty.default)(data)) {
        throw new Error(`No description found for table ${table.tableName}${table.schema ? ` in schema ${table.schema}` : ""}. Check the table name and schema; remember, they _are_ case sensitive.`);
      }
      return data;
    } catch (error) {
      if (error instanceof import_errors.BaseError && ((_a = error.cause) == null ? void 0 : _a.code) === "ER_NO_SUCH_TABLE") {
        throw new Error(`No description found for table ${table.tableName}${table.schema ? ` in schema ${table.schema}` : ""}. Check the table name and schema; remember, they _are_ case sensitive.`);
      }
      throw error;
    }
  }
  /**
   * Removes a column from a table
   *
   * @param tableName
   * @param columnName
   * @param options
   */
  async removeColumn(tableName, columnName, options) {
    const queryOptions = { ...options, raw: true };
    const sql = this.queryGenerator.removeColumnQuery(tableName, columnName, queryOptions);
    await this.sequelize.queryRaw(sql, queryOptions);
  }
  /**
   * Add a constraint to a table
   *
   * Available constraints:
   * - UNIQUE
   * - DEFAULT (MSSQL only)
   * - CHECK (Not supported by MySQL)
   * - FOREIGN KEY
   * - PRIMARY KEY
   *
   * @example UNIQUE
   * ```ts
   * queryInterface.addConstraint('Users', {
   *   fields: ['email'],
   *   type: 'UNIQUE',
   *   name: 'custom_unique_constraint_name'
   * });
   * ```
   *
   * @example CHECK
   * ```ts
   * queryInterface.addConstraint('Users', {
   *   fields: ['roles'],
   *   type: 'CHECK',
   *   where: {
   *      roles: ['user', 'admin', 'moderator', 'guest']
   *   }
   * });
   * ```
   *
   * @example Default - MSSQL only
   * ```ts
   * queryInterface.addConstraint('Users', {
   *    fields: ['roles'],
   *    type: 'DEFAULT',
   *    defaultValue: 'guest'
   * });
   * ```
   *
   * @example Primary Key
   * ```ts
   * queryInterface.addConstraint('Users', {
   *    fields: ['username'],
   *    type: 'PRIMARY KEY',
   *    name: 'custom_primary_constraint_name'
   * });
   * ```
   *
   * @example Composite Primary Key
   * ```ts
   * queryInterface.addConstraint('Users', {
   *    fields: ['first_name', 'last_name'],
   *    type: 'PRIMARY KEY',
   *    name: 'custom_primary_constraint_name'
   * });
   * ```
   *
   * @example Foreign Key
   * ```ts
   * queryInterface.addConstraint('Posts', {
   *   fields: ['username'],
   *   type: 'FOREIGN KEY',
   *   name: 'custom_fkey_constraint_name',
   *   references: { //Required field
   *     table: 'target_table_name',
   *     field: 'target_column_name'
   *   },
   *   onDelete: 'cascade',
   *   onUpdate: 'cascade'
   * });
   * ```
   *
   * @example Composite Foreign Key
   * ```ts
   * queryInterface.addConstraint('TableName', {
   *   fields: ['source_column_name', 'other_source_column_name'],
   *   type: 'FOREIGN KEY',
   *   name: 'custom_fkey_constraint_name',
   *   references: { //Required field
   *     table: 'target_table_name',
   *     fields: ['target_column_name', 'other_target_column_name']
   *   },
   *   onDelete: 'cascade',
   *   onUpdate: 'cascade'
   * });
   * ```
   *
   * @param tableName - Table name where you want to add a constraint
   * @param options - An object to define the constraint name, type etc
   */
  async addConstraint(tableName, options) {
    if (!options.fields) {
      throw new Error("Fields must be specified through options.fields");
    }
    if (!options.type) {
      throw new Error("Constraint type must be specified through options.type");
    }
    const sql = this.queryGenerator.addConstraintQuery(tableName, options);
    await this.sequelize.queryRaw(sql, { ...options, raw: true, type: import_query_types.QueryTypes.RAW });
  }
  async deferConstraints(constraintChecking, options) {
    (0, import_model_internals.setTransactionFromCls)(options ?? {}, this.sequelize);
    if (!(options == null ? void 0 : options.transaction)) {
      throw new Error("Missing transaction in deferConstraints option.");
    }
    const sql = this.queryGenerator.setConstraintCheckingQuery(constraintChecking);
    await this.sequelize.queryRaw(sql, { ...options, raw: true, type: import_query_types.QueryTypes.RAW });
  }
  /**
   * Remove a constraint from a table
   *
   * @param tableName -Table name to drop constraint from
   * @param constraintName -Constraint name
   * @param options -Query options
   */
  async removeConstraint(tableName, constraintName, options) {
    const sql = this.queryGenerator.removeConstraintQuery(tableName, constraintName, options);
    await this.sequelize.queryRaw(sql, { ...options, raw: true, type: import_query_types.QueryTypes.RAW });
  }
  async showConstraints(tableName, options) {
    const sql = this.queryGenerator.showConstraintsQuery(tableName, options);
    const rawConstraints = await this.sequelize.queryRaw(sql, { ...options, raw: true, type: import_query_types.QueryTypes.SHOWCONSTRAINTS });
    const constraintMap = /* @__PURE__ */ new Map();
    for (const {
      columnNames,
      definition,
      deleteAction,
      initiallyDeferred,
      isDeferrable,
      referencedColumnNames,
      referencedTableName,
      referencedTableSchema,
      updateAction,
      ...rawConstraint
    } of rawConstraints) {
      const constraint = constraintMap.get(rawConstraint.constraintName);
      if (constraint) {
        if (columnNames) {
          constraint.columnNames = constraint.columnNames ? [.../* @__PURE__ */ new Set([...constraint.columnNames, columnNames])] : [columnNames];
        }
        if (referencedColumnNames) {
          constraint.referencedColumnNames = constraint.referencedColumnNames ? [.../* @__PURE__ */ new Set([...constraint.referencedColumnNames, referencedColumnNames])] : [referencedColumnNames];
        }
      } else {
        const constraintData = { ...rawConstraint };
        if (columnNames) {
          constraintData.columnNames = [columnNames];
        }
        if (referencedTableSchema) {
          constraintData.referencedTableSchema = referencedTableSchema;
        }
        if (referencedTableName) {
          constraintData.referencedTableName = referencedTableName;
        }
        if (referencedColumnNames) {
          constraintData.referencedColumnNames = [referencedColumnNames];
        }
        if (deleteAction) {
          constraintData.deleteAction = deleteAction.replaceAll("_", " ");
        }
        if (updateAction) {
          constraintData.updateAction = updateAction.replaceAll("_", " ");
        }
        if (definition) {
          constraintData.definition = definition;
        }
        if (this.sequelize.dialect.supports.constraints.deferrable) {
          constraintData.deferrable = isDeferrable ? initiallyDeferred === "YES" ? import_deferrable.Deferrable.INITIALLY_DEFERRED : import_deferrable.Deferrable.INITIALLY_IMMEDIATE : import_deferrable.Deferrable.NOT;
        }
        constraintMap.set(rawConstraint.constraintName, constraintData);
      }
    }
    return [...constraintMap.values()];
  }
  /**
   * Returns all foreign key constraints of requested tables
   *
   * @deprecated Use {@link showConstraints} instead.
   * @param _tableNames
   * @param _options
   */
  getForeignKeysForTables(_tableNames, _options) {
    throw new Error(`getForeignKeysForTables has been deprecated. Use showConstraints instead.`);
  }
  /**
   * Get foreign key references details for the table
   *
   * @deprecated Use {@link showConstraints} instead.
   * @param _tableName
   * @param _options
   */
  getForeignKeyReferencesForTable(_tableName, _options) {
    throw new Error(`getForeignKeyReferencesForTable has been deprecated. Use showConstraints instead.`);
  }
  async withoutForeignKeyChecks(optionsOrCallback, maybeCallback) {
    let options;
    let callback;
    if (typeof optionsOrCallback === "function") {
      options = {};
      callback = optionsOrCallback;
    } else {
      options = { ...optionsOrCallback };
      callback = maybeCallback;
    }
    (0, import_model_internals.setTransactionFromCls)(options, this.sequelize);
    if (options.connection) {
      return this.#withoutForeignKeyChecks(options, callback);
    }
    return this.sequelize.withConnection(async (connection) => {
      return this.#withoutForeignKeyChecks({ ...options, connection }, callback);
    });
  }
  async #withoutForeignKeyChecks(options, cb) {
    try {
      await this.unsafeToggleForeignKeyChecks(false, options);
      return await cb(options.connection);
    } finally {
      await this.unsafeToggleForeignKeyChecks(true, options);
    }
  }
  /**
   * Toggles foreign key checks.
   * Don't forget to turn them back on, use {@link withoutForeignKeyChecks} to do this automatically.
   *
   * @param enable
   * @param options
   */
  async unsafeToggleForeignKeyChecks(enable, options) {
    await this.sequelize.queryRaw(this.queryGenerator.getToggleForeignKeyChecksQuery(enable), options);
  }
}
